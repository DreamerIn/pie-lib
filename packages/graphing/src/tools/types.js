import PropTypes from 'prop-types';

export const ToolPropType = {
  mark: PropTypes.any,
  onChange: PropTypes.func,
  onDragStart: PropTypes.func,
  onDragStop: PropTypes.func
};

export const PointType = {
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired
};
