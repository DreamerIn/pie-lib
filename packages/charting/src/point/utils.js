import debug from 'debug';
import omitBy from 'lodash/omitBy';
import isUndefined from 'lodash/isUndefined';

const log = debug('pie-lib:charting:point:utils');

/**
 * Get the index of the point in the points array
 * @param {Point[]} points
 * @param {Point} p
 */
export const pointIndex = (points, p) => {
  return (points || []).findIndex(s => {
    return s.x === p.x && s.y === p.y;
  });
};

export const hasPoint = (points, p) => pointIndex(points, p) !== -1;

export const trim = point => ({
  x: point.x,
  y: point.y,
  label: point.label
});

export const removePoints = (points, toRemove) => {
  log('[removePoints] points: ', points, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];

  const out = points.filter(p => !hasPoint(toRemove, p));

  log('[removePoints] points: ', out);
  return out;
};

export const firstAvailableLabel = (points, labels) => {
  const existing = points.map(p => p.label);
  log('points: ', points);
  const filtered = labels.filter(l => !existing.some(e => e === l));
  log('existing: ', existing, 'filtered: ', filtered);
  return filtered.length > 0 ? filtered[0] : '';
};

export const addPoint = (points, p, labels) => {
  labels = labels || [];
  if (!hasPoint(points, p)) {
    const out = points.slice();
    const label = firstAvailableLabel(points, labels);
    out.push(Object.assign({}, p, { label }));
    return out;
  } else {
    return points;
  }
};

export const bounds = (p, domain, range) => {
  return {
    left: domain.min - p.x,
    right: Math.abs(p.x - domain.max),
    top: Math.abs(p.y - range.max),
    bottom: range.min - p.y
  };
};

export const swap = (arr, from, to, getIndex) => {
  log('[swap] arr: ', arr, 'from: ', from, 'to: ', to);

  const toIndex = getIndex(arr, to);

  log('[swap] toIndex: ', toIndex);

  if (toIndex !== -1) {
    //prevent moving one point on top of another.
    return arr;
  }

  const fromIndex = getIndex(arr, from);

  log('[swap] fromIndex: ', fromIndex);

  if (fromIndex === -1) {
    //cant move a point that isn't in the points array
    return arr;
  }

  const f = arr[fromIndex];
  const replacement = Object.assign({}, f, omitBy(to, isUndefined));
  const out = arr.slice();
  out.splice(fromIndex, 1, replacement); //{ ...to, label: f.label });
  return out;
};

/**
 * swap a point 'from', with point 'to'.
 * Finds the point in the array and updates it to equal the 'to' point.
 * @param {Point[]} points
 * @param {Point} from
 * @param {Point} to
 */

export const swapPoint = (points, from, to) => {
  return swap(points, from, to, pointIndex);
};
