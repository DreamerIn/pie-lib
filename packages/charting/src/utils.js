import { scaleBand, scalePoint } from '@vx/scale';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width, type) => {
  switch (type) {
    case 'bar':
    case 'dotPlot':
    case 'linePlot':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0.2
      });
    case 'histogram':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
    case 'line':
      return scalePoint({
        domain: data && data.map(bandKey),
        rangeRound: [0, width]
      });
    default:
      return scaleBand({
        range: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
  }
};

export const getTickValues = (prop = {}) => {
  const tickValues = [];
  let tickVal = prop.min;

  while (tickVal <= prop.max) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 100) / 100;
  }

  return tickValues;
};

export const getDomainAndRangeByChartType = (domain, range, chartType) => {
  const { min, max } = range || {};
  let { step } = range || {};
  const numberMax = parseFloat(max);

  if (!step) {
    if (numberMax < 20) {
      step = 1;
    } else if (numberMax >= 20 && numberMax < 100) {
      step = 5;
    } else {
      step = 10;
    }
  }

  switch (chartType) {
    // if chart is dot plot or line plot, we should ignore step and make sure that min & max are integer values
    case 'dotPlot':
    case 'linePlot':
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          min: parseInt(min, 10),
          max: parseInt(max, 10),
          step: 1
        }
      };
    default:
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          step
        }
      };
  }
};

export const getGridLinesAndAxisByChartType = (range, chartType) => {
  switch (chartType) {
    case 'line':
      return {
        verticalLines: undefined,
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
    case 'dotPlot':
    case 'linePlot':
      return {
        verticalLines: [],
        horizontalLines: [],
        leftAxis: false
      };
    default:
      return {
        verticalLines: [],
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
  }
};

export const getRotateAngle = barWidth => {
  if (barWidth < 30) {
    return 75;
  }

  if (barWidth < 40) {
    return 45;
  }

  if (barWidth < 60) {
    return 25;
  }

  return 0;
};

export const getTopPadding = barWidth => {
  if (barWidth < 30) {
    return 50;
  }

  if (barWidth < 40) {
    return 30;
  }

  if (barWidth < 60) {
    return 15;
  }

  return 0;
};
