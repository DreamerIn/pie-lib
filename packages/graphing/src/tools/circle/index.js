import Circle from './component';
import isEqual from 'lodash/isEqual';

export const tool = () => ({
  type: 'circle',
  Component: Circle,
  hover: (point, mark) => {
    return { ...mark, edge: point };
  },
  addPoint: (point, mark) => {
    if (mark && isEqual(mark.root, point)) {
      return mark;
    }

    if (!mark) {
      return {
        type: 'circle',
        root: point,
        building: true
      };
    } else {
      return { ...mark, edge: point, building: false };
    }
  }
});
