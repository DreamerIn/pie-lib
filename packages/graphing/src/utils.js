import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

// export { * from utils}
export const polygonToArea = points => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => {
  const left = Math.min(from.x, to.x);
  const top = Math.max(from.y, to.y);
  const bottom = Math.min(from.y, to.y);
  const right = Math.max(from.x, to.x);
  return { left, top, bottom, right };
};
