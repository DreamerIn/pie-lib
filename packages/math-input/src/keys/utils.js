export const mkSet = category => o => ({ ...o, category });

export const toArray = o => Object.keys(o).map(k => o[k]);

export const toKey = category => v => {
  if (typeof v === 'string') {
    return {
      name: v,
      label: v,
      category,
      write: v
    };
  } else {
    return {
      ...v,
      category
    };
  }
};
