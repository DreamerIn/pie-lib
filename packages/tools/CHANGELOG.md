# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.4.0"></a>
# [0.4.0](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.3.0...@pie-lib/tools@0.4.0) (2018-05-01)


### Features

* **ruler:** move handles away from ticks ([85353fb](https://github.com/pie-framework/pie-lib/commit/85353fb))




<a name="0.3.0"></a>
# [0.3.0](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.2.2...@pie-lib/tools@0.3.0) (2018-05-01)


### Features

* **ruler:** add className and startPosition ([bca0544](https://github.com/pie-framework/pie-lib/commit/bca0544))




<a name="0.2.2"></a>
## [0.2.2](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.2.1...@pie-lib/tools@0.2.2) (2018-04-30)


### Bug Fixes

* **protractor:** move lines out of circle ([505178d](https://github.com/pie-framework/pie-lib/commit/505178d))




<a name="0.2.1"></a>
## [0.2.1](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.2.0...@pie-lib/tools@0.2.1) (2018-04-30)


### Bug Fixes

* **protractor:** make center circle transparent ([e748293](https://github.com/pie-framework/pie-lib/commit/e748293))




<a name="0.2.0"></a>
# [0.2.0](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.1.3...@pie-lib/tools@0.2.0) (2018-04-30)


### Features

* **rotatable:** add startPosition and className ([4e93991](https://github.com/pie-framework/pie-lib/commit/4e93991))




<a name="0.1.3"></a>
## [0.1.3](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.1.2...@pie-lib/tools@0.1.3) (2018-04-30)


### Bug Fixes

* **dependencies:** set material-ui version to ^1.0.0-beta.44 ([6fd66bc](https://github.com/pie-framework/pie-lib/commit/6fd66bc))




<a name="0.1.2"></a>
## [0.1.2](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.1.1...@pie-lib/tools@0.1.2) (2018-04-30)


### Bug Fixes

* **dependencies:** lock material-ui to 1.0.0-beta.44 ([60df8e6](https://github.com/pie-framework/pie-lib/commit/60df8e6))




<a name="0.1.1"></a>
## [0.1.1](https://github.com/pie-framework/pie-lib/compare/@pie-lib/tools@0.1.0...@pie-lib/tools@0.1.1) (2018-04-30)


### Bug Fixes

* **dependencies:** lock material-ui to 1.0.0-beta.43 ([2a3e087](https://github.com/pie-framework/pie-lib/commit/2a3e087))
* **dependencies:** upgrade material-ui ([b94b50e](https://github.com/pie-framework/pie-lib/commit/b94b50e))




<a name="0.1.0"></a>
# 0.1.0 (2018-04-30)


### Features

* **tools:** Add new 'tools' package. ([9217d0c](https://github.com/pie-framework/pie-lib/commit/9217d0c))
