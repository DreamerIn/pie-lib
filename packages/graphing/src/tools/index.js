import { tool as point } from './point';
import { tool as circle } from './circle';
import { tool as polygon } from './polygon';

export { point, circle, polygon };
