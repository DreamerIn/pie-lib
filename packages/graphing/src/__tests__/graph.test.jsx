import { shallow } from 'enzyme';
import React from 'react';

import { xy } from '../__tests__/utils';

import Graph, { removeBuildingToolIfCurrentToolDiffers } from '../graph';
import { toolsArr } from '../tools';

const marks = [
  {
    type: 'point',
    x: 2,
    y: 2,
    label: 'Point',
    showLabel: true
  },
  {
    type: 'line',
    from: { x: 0, y: 0 },
    label: 'Line',
    building: true
  },
];

describe('removeBuildingToolIfCurrentToolDiffers', () => {
  it('keeps all marks if currentTool is the same', () => {
    const currentTool = toolsArr[0];

    expect(removeBuildingToolIfCurrentToolDiffers({ marks, currentTool, state: { currentTool } })).toEqual(marks)
  });

  it('removes building marks if currentTool is different', () => {
    const currentTool = toolsArr[0];
    const stateCurrentTool = toolsArr[1];

    expect(removeBuildingToolIfCurrentToolDiffers({
      marks: [...marks],
      currentTool,
      state: { currentTool: stateCurrentTool }
    })).toEqual([marks[0]]);
  });
});

describe('Graph', () => {
  let w;
  let onChangeMarks = jest.fn();
  const defaultProps = () => ({
    classes: {},
    className: 'className',
    onChangeMarks,
    tools: toolsArr,
    domain: { min: 0, max: 1, step: 1 },
    range: { min: 0, max: 1, step: 1 },
    size: { width: 400, height: 400 },
    marks,
    currentTool: toolsArr[0]
  });
  const initialProps = defaultProps();

  const wrapper = (extras, opts) => {
    const props = { ...initialProps, ...extras };

    return shallow(<Graph {...props} />, opts);
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('getDerivedStateFromProps', () => {
      it('removes the building mark if currentTool is different', () => {
        const result = Graph.getDerivedStateFromProps(initialProps);

        expect(result.marks).toEqual([marks[0]]);
      });

      it('keeps all marks if currentTool is the same', () => {
        const result = Graph.getDerivedStateFromProps(initialProps, { currentTool: initialProps.currentTool });

        expect(result.marks).toEqual(marks);
      });
    });

    describe('componentDidMount', () => {
      it('sets the labelNode to state', () => {
        w = shallow(<Graph {...defaultProps()} />, { disableLifecycleMethods: true });
        w.instance().labelNode = {};
        w.instance().componentDidMount();
        expect(w.state('labelNode')).toEqual(w.instance().labelNode);
      });
    });

    describe('changeMark', () => {
      const newMark = { type: 'mark', x: 2, y: 2 };

      it('does not call onChangeMarks', () => {
        w = wrapper();
        w.instance().changeMark(newMark, newMark);
        expect(onChangeMarks).not.toBeCalled();
      });

      it('calls onChangeMarks', () => {
        w = wrapper();
        w.instance().changeMark(marks[0], newMark);
        expect(onChangeMarks).toHaveBeenCalledWith([newMark, marks[1]]);
      });
    });

    describe('completeMark', () => {
      it('calls updateMarks', () => {
        const completedMark = { ...marks[1], completed: true };

        const complete = jest.fn().mockReturnValue(completedMark);
        const updateMarks = jest.fn();

        w.instance().state.marks = [...marks];
        w.instance().state.currentTool.complete = complete;
        w.instance().updateMarks = updateMarks;
        w.instance().completeMark({ x: 3, y: 3 });

        expect(complete).toHaveBeenCalledWith(marks[1], xy(3, 3));
        expect(updateMarks).toHaveBeenCalledWith(marks[1], completedMark);
      });

      it('does not call updateMarks if no building mark', () => {
        const completedMark = { type: 'mark', completed: true };

        const complete = jest.fn().mockReturnValue(completedMark);
        const updateMarks = jest.fn();

        w.instance().state.marks = [marks[0]];
        w.instance().state.currentTool.complete = complete;
        w.instance().updateMarks = updateMarks;
        w.instance().completeMark({ x: 3, y: 3 });

        expect(complete).not.toBeCalled();
        expect(updateMarks).not.toBeCalled();
      });

      it('does not call updateMarks if no current tool', () => {
        const completedMark = { type: 'mark', completed: true };

        const complete = jest.fn().mockReturnValue(completedMark);
        const updateMarks = jest.fn();

        w.instance().state.currentTool = null;
        w.instance().state.marks = [...marks];
        w.instance().updateMarks = updateMarks;
        w.instance().completeMark({ x: 3, y: 3 });

        expect(complete).not.toBeCalled();
        expect(updateMarks).not.toBeCalled();
      });
    });

    describe('updateMarks', () => {
      it('calls onChangeMarks', () => {
        const marks = [{ type: 'mark', ...xy(2, 2) }];
        const update = { type: 'mark', ...xy(4, 4) };
        w = wrapper({ marks });
        w.instance().updateMarks(marks[0], update, false);
      });

      it('calls onChangeMarks with added mark', () => {
        const marks = [];
        const update = { type: 'mark', ...xy(4, 4) };
        w = wrapper({ marks });
        w.instance().updateMarks(marks[0], [update], true);
      });
    });

    describe('getComponent', () => {
      let compMock = jest.fn();

      it('returns null if no mark', () => {
        expect(w.instance().getComponent()).toEqual(null);
        expect(w.instance().getComponent(undefined)).toEqual(null);
        expect(w.instance().getComponent(null)).toEqual(null);
      });

      it('returns the component', () => {
        w.instance().props.tools[0].Component = compMock;

        const Comp = w.instance().getComponent({ type: toolsArr[0].type });
        expect(Comp).toEqual(compMock);
      });

      it('returns null if there is no tool', () => {
        expect(w.instance().getComponent({ type: 'mark' })).toEqual(null);
      });

      it('returns null if there is no tool.Component', () => {
        w.instance().props.tools[0].Component = undefined;

        expect(w.instance().getComponent({ type: toolsArr[0].type })).toEqual(null);
      });
    });

    describe('mouseMove', () => {
      it('updates state if there is a buildingMark and !dragging', () => {
        const marks = [{ type: 'mark' }];
        w = wrapper({ marks });
        w.setState({ buildingMark: { type: 'mark' } });
        w.instance().mouseMove(xy(3, 3));
        expect(w.state().hoverPoint).toEqual(xy(3, 3));
      });

      it('does not update state if there is no buildingMark', () => {
        const marks = [{ type: 'mark' }];
        w = wrapper({ marks });
        w.instance().mouseMove(xy(3, 3));
        expect(w.state().hoverPoint).toBeUndefined();
      });

      it('does not update state if dragging', () => {
        const marks = [{ type: 'mark' }];
        w = wrapper({ marks });
        w.setState({ dragging: true });
        w.instance().mouseMove(xy(3, 3));
        expect(w.state().hoverPoint).toBeUndefined();
      });
    });

    describe('onBgClick', () => {
      it('calls updateMarks', () => {
        const buildingMark = { type: 'mark', building: true, x: 1, y: 1 };

        const updatedMark = { type: 'mark', msg: 'updated' };
        const marks = [{ type: 'mark' }, buildingMark];

        const addPoint = jest.fn().mockReturnValue(updatedMark);
        const updateMarks = jest.fn();

        w = wrapper({ marks });
        w.instance().state.currentTool.addPoint = addPoint;
        w.instance().updateMarks = updateMarks;
        w.instance().onBgClick({ x: 3, y: 3 });
        expect(w.instance().updateMarks).toHaveBeenCalledWith(buildingMark, updatedMark, true);
      });

      it('returns early of labelModeEnabled', () => {
        w = wrapper({ labelModeEnabled: true });
        w.instance().updateMarks = jest.fn();
        w.instance().onBgClick({ x: 3, y: 3 });
        expect(w.instance().updateMarks).not.toHaveBeenCalled();
      });
    });
  });
});
