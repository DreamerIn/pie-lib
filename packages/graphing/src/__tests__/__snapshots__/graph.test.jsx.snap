// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph snapshot renders 1`] = `
<WithStyles(Root)
  graphProps={
    Object {
      "domain": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "range": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "scale": Object {
        "x": [Function],
        "y": [Function],
      },
      "size": Object {
        "height": 400,
        "width": 400,
      },
      "snap": Object {
        "x": [Function],
        "y": [Function],
      },
    }
  }
  onMouseMove={[Function]}
>
  <WithStyles(Grid)
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [Function],
          "y": [Function],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [Function],
          "y": [Function],
        },
      }
    }
  />
  <Axes
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [Function],
          "y": [Function],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [Function],
          "y": [Function],
        },
      }
    }
    includeArrows={true}
  />
  <Bg
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [Function],
          "y": [Function],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [Function],
          "y": [Function],
        },
      }
    }
    height={400}
    onClick={[Function]}
    width={400}
  />
  <WithStyles(Labels)
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [Function],
          "y": [Function],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [Function],
          "y": [Function],
        },
      }
    }
  />
</WithStyles(Root)>
`;
